lib/storage/src/content_manager/consensus_manager.rs

```diff
@@ -788,6 +802,87 @@ impl<C: CollectionContainer> ConsensusManager<C> {
     }
 }
 
+fn recover_first_voter(
+    wal: &ConsensusOpWal,
+    peers: &[PeerId],
+) -> Result<Option<PeerId>, StorageError> {
+    let Some(first_entry) = wal.first_entry()? else {
+        log::debug!("Skipped recovering first voter peer: WAL is empty");
+        return Ok(None);
+    };
+
+    let Some(last_entry) = wal.last_entry()? else {
+        log::error!(
+            "Failed to recover first voter peer: \
+             WAL contains first entry, but no last entry"
+        );
+
+        return Ok(None);
+    };
+
+    if first_entry.index != 1 {
+        log::warn!("Failed to recover first voter peer: WAL is truncated");
+        return Ok(Some(PeerId::MAX));
+    }
+
+    // Try to recover first voter peer from WAL (if it was not removed from cluster yet!):
+    // - collect a list of current peers
+    // - scroll WAL and *remove* a peer from the list when `AddPeer`/`AddLearnerPeer` operation encountered
+    // - if there's exactly one peer left in the list at the end, this peer should be the first voter
+
+    let mut peers: HashSet<_> = peers.iter().copied().collect();
+
+    for index in first_entry.index..last_entry.index + 1 {
+        let entry = wal.entry(index)?;
+
+        match entry.get_entry_type() {
+            EntryType::EntryConfChangeV2 => {
+                let change: ConfChangeV2 = prost_for_raft::Message::decode(entry.get_data())?;
+
+                for change in change.changes {
+                    match change.get_change_type() {
+                        ConfChangeType::AddNode | ConfChangeType::AddLearnerNode => {
+                            peers.remove(&change.get_node_id());
+                        }
+
+                        ConfChangeType::RemoveNode => (),
+                    }
+                }
+            }
+
+            EntryType::EntryConfChange => {
+                log::warn!(
+                    "Encountered deprecated ConfChange message while recovering first voter peer"
+                );
+
+                let change: ConfChange = prost_for_raft::Message::decode(entry.get_data())?;
+
+                match change.get_change_type() {
+                    ConfChangeType::AddNode | ConfChangeType::AddLearnerNode => {
+                        peers.remove(&change.get_node_id());
+                    }
+
+                    ConfChangeType::RemoveNode => (),
+                }
+            }
+
+            EntryType::EntryNormal => continue,
+        }
+    }
+
+    if peers.len() > 1 {
+        log::warn!(
+            "Failed to recover first voter peer: \
+             found multiple peers without ConfChange entry in WAL: \
+             {peers:?}"
+        );
+
+        return Ok(Some(PeerId::MAX));
+    }
```

Comment:
Also check and if the list is empty?

I think we can report:

> Failed to recover first voter peer: first peer has already been removed from the cluster

Reply:
If `peers` is empty by the end of the test, it means first peer **is** in the WAL. It's "handled" by the `peers.into_iter().next()` (which will return `None`) at the end of the method.

Though, maybe I should change it to return `PeerId::MAX`, so we don't try `recover_first_voter` again.

Reply:
> it means first peer is in the WAL. It's "handled" by the peers.into_iter().next()

Ah of course, in cases where we did add the origin peer as first entry.